// @remotion/studio-server/dist/preview-server/routes/insert-element.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.insertElementHandler = void 0;
const node_fs_1 = require("node:fs");
const node_path_1 = __importDefault(require("node:path"));
const renderer_1 = require("@remotion/renderer");
const studio_shared_1 = require("@remotion/studio-shared");
const file_watcher_1 = require("../../file-watcher");
const resolve_composition_component_1 = require("../../helpers/resolve-composition-component");
const format_log_file_location_1 = require("../format-log-file-location");
const undo_stack_1 = require("../undo-stack");
const log_update_1 = require("./log-updates/log-update");
const save_props_mutex_1 = require("./save-props-mutex");
const validatePosition = (position) => {
    if (position === null) {
        return;
    }
    if (!Number.isFinite(position.x) || !Number.isFinite(position.y)) {
        throw new Error('Position must be finite');
    }
};
const isInside = ({ child, parent }) => {
    const relative = node_path_1.default.relative(parent, child);
    return (relative === '' ||
        (!relative.startsWith('..') && !node_path_1.default.isAbsolute(relative)));
};
const withoutTsxExtension = (fileName) => {
    return fileName.replace(/\.tsx$/, '');
};
const normalizeSourceForComparison = (source) => {
    return source.replace(/\r\n/g, '\n').trim();
};
const makeRelativeImportPath = ({ fromFile, toFile, }) => {
    const withoutExtension = withoutTsxExtension(toFile);
    let relative = node_path_1.default
        .relative(node_path_1.default.dirname(fromFile), withoutExtension)
        .split(node_path_1.default.sep)
        .join('/');
    if (!relative.startsWith('.')) {
        relative = `./${relative}`;
    }
    return relative;
};
const validateDimensions = (dimensions) => {
    if (!Number.isFinite(dimensions.width) ||
        !Number.isFinite(dimensions.height) ||
        dimensions.width <= 0 ||
        dimensions.height <= 0) {
        throw new Error('Element dimensions must be positive finite numbers');
    }
};
const validateElement = (element) => {
    if ((0, studio_shared_1.makeElementFileNameFromSlug)(element.slug) === null) {
        throw new Error('Element slug must produce a safe lowercase .tsx file name');
    }
    if (typeof element.sourceCode !== 'string' ||
        element.sourceCode.trim().length === 0 ||
        element.sourceCode.length > 200000) {
        throw new Error('Unsupported Element source code');
    }
    if ((0, studio_shared_1.getElementComponentNameFromSourceCode)(element.sourceCode) === null) {
        throw new Error('Element source must export exactly one named component');
    }
    validateDimensions(element.dimensions);
};
const insertElementHandler = ({ input: { compositionFile, compositionId, element, position }, remotionRoot, logLevel, }) => (0, save_props_mutex_1.withSavePropsLock)(async () => {
    try {
        validateElement(element);
        validatePosition(position);
        const componentName = (0, studio_shared_1.getElementComponentNameFromSourceCode)(element.sourceCode);
        if (componentName === null) {
            throw new Error('Element source must export exactly one named component');
        }
        renderer_1.RenderInternals.Log.trace({ indent: false, logLevel }, `[insert-element] Received request for compositionFile="${compositionFile}" compositionId="${compositionId}" element="${element.slug}"`);
        const location = await (0, resolve_composition_component_1.resolveCompositionComponentWithFile)({
            remotionRoot,
            compositionFile,
            compositionId,
        });
        if (!location.canAddSequence) {
            throw new Error('Cannot insert Element into this composition component');
        }
        const derivedElementFileName = (0, studio_shared_1.makeElementFileNameFromSlug)(element.slug);
        if (derivedElementFileName === null) {
            throw new Error('Element slug must produce a safe lowercase .tsx file name');
        }
        const elementFileName = node_path_1.default.resolve(node_path_1.default.dirname(location.fileName), derivedElementFileName);
        if (!isInside({ child: elementFileName, parent: remotionRoot })) {
            throw new Error('Element file must stay inside the Remotion project');
        }
        const elementFileExists = (0, node_fs_1.existsSync)(elementFileName);
        if (elementFileExists) {
            const existingSource = (0, node_fs_1.readFileSync)(elementFileName, 'utf-8');
            if (normalizeSourceForComparison(existingSource) !==
                normalizeSourceForComparison(element.sourceCode)) {
                throw new Error(`Element file already exists with different contents: ${derivedElementFileName}`);
            }
        }
        const importPath = makeRelativeImportPath({
            fromFile: location.fileName,
            toFile: elementFileName,
        });
        const inserted = await (0, resolve_composition_component_1.insertJsxElementIntoComposition)({
            remotionRoot,
            compositionFile,
            compositionId,
            element: {
                type: 'component',
                componentName,
                importName: componentName,
                importPath,
                props: [],
                position: null,
            },
            prettierConfigOverride: null,
            wrapInSequence: {
                dimensions: element.dimensions,
                name: element.displayName,
                position,
            },
        });
        (0, undo_stack_1.pushTransactionToUndoStack)({
            snapshots: [
                ...(elementFileExists
                    ? []
                    : [
                        {
                            filePath: elementFileName,
                            oldContents: null,
                            newContents: element.sourceCode,
                            logLine: 1,
                        },
                    ]),
                {
                    filePath: inserted.fileName,
                    oldContents: inserted.oldContents,
                    newContents: inserted.output,
                    logLine: inserted.logLine,
                },
            ],
            logLevel,
            remotionRoot,
            description: {
                undoMessage: `↩️  Added ${element.displayName}`,
                redoMessage: `↪️  Added ${element.displayName}`,
            },
            entryType: 'insert-jsx-element',
            suppressHmrOnFileRestore: false,
        });
        if (!elementFileExists) {
            (0, undo_stack_1.suppressUndoStackInvalidation)(elementFileName);
        }
        (0, undo_stack_1.suppressUndoStackInvalidation)(inserted.fileName);
        if (!elementFileExists) {
            (0, file_watcher_1.writeFileAndNotifyFileWatchers)(elementFileName, element.sourceCode, undefined);
        }
        (0, file_watcher_1.writeFileAndNotifyFileWatchers)(inserted.fileName, inserted.output, undefined);
        const compositionLocationLabel = (0, format_log_file_location_1.formatLogFileLocation)({
            remotionRoot,
            absolutePath: inserted.fileName,
            line: inserted.logLine,
        });
        const elementLocationLabel = (0, format_log_file_location_1.formatLogFileLocation)({
            remotionRoot,
            absolutePath: elementFileName,
            line: 1,
        });
        renderer_1.RenderInternals.Log.info({ indent: false, logLevel }, `${renderer_1.RenderInternals.chalk.blueBright(elementLocationLabel)} ${elementFileExists ? 'Reused existing Element source' : 'Created Element source'}`);
        renderer_1.RenderInternals.Log.info({ indent: false, logLevel }, `${renderer_1.RenderInternals.chalk.blueBright(compositionLocationLabel)} Added <${componentName}>`);
        if (!inserted.formatted) {
            (0, log_update_1.warnAboutPrettierOnce)(logLevel);
        }
        (0, undo_stack_1.printUndoHint)(logLevel);
        return {
            success: true,
        };
    }
    catch (err) {
        return {
            success: false,
            reason: err.message,
            stack: err.stack,
        };
    }
});
exports.insertElementHandler = insertElementHandler;
