// @remotion/studio-shared/dist/element-drag-data.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.parseElementDragData = exports.makeElementDragData = exports.getElementComponentNameFromSourceCode = exports.makeElementFileNameFromSlug = exports.isLowercaseElementFileName = exports.ELEMENT_DRAG_MIME_TYPE = void 0;
const component_drag_data_1 = require("./component-drag-data");
exports.ELEMENT_DRAG_MIME_TYPE = 'application/vnd.remotion.element+json';
const isRecord = (value) => {
    return typeof value === 'object' && value !== null && !Array.isArray(value);
};
const isLowercaseElementFileName = (value) => {
    return (typeof value === 'string' &&
        value.length > 0 &&
        value.length < 120 &&
        value === value.toLowerCase() &&
        value.endsWith('.tsx') &&
        !value.includes('/') &&
        !value.includes('\\') &&
        !value.includes('\0') &&
        !value.includes('..') &&
        /^[a-z0-9][a-z0-9.-]*\.tsx$/.test(value));
};
exports.isLowercaseElementFileName = isLowercaseElementFileName;
const isSlug = (value) => {
    return (typeof value === 'string' &&
        value.length > 0 &&
        value.length < 120 &&
        /^[a-z0-9][a-z0-9/-]*$/.test(value) &&
        !value.includes('..') &&
        !value.includes('//'));
};
const makeElementFileNameFromSlug = (slug) => {
    if (!isSlug(slug)) {
        return null;
    }
    const lastSegment = slug.split('/').at(-1);
    if (!lastSegment) {
        return null;
    }
    const fileName = `${lastSegment}.element.tsx`;
    return (0, exports.isLowercaseElementFileName)(fileName) ? fileName : null;
};
exports.makeElementFileNameFromSlug = makeElementFileNameFromSlug;
const isDisplayName = (value) => {
    return typeof value === 'string' && value.length > 0 && value.length < 120;
};
const isSourceCode = (value) => {
    return (typeof value === 'string' &&
        value.trim().length > 0 &&
        value.length < 200000);
};
const isDimensions = (value) => {
    if (!isRecord(value)) {
        return false;
    }
    return (typeof value.width === 'number' &&
        Number.isFinite(value.width) &&
        value.width > 0 &&
        typeof value.height === 'number' &&
        Number.isFinite(value.height) &&
        value.height > 0);
};
const getElementComponentNameFromSourceCode = (sourceCode) => {
    const componentNames = Array.from(sourceCode.matchAll(/export\s+(?:const|function)\s+([A-Z_$][A-Za-z0-9_$]*)\b/g)).map((match) => match[1]);
    const uniqueComponentNames = Array.from(new Set(componentNames));
    if (uniqueComponentNames.length !== 1) {
        return null;
    }
    const [componentName] = uniqueComponentNames;
    return (0, component_drag_data_1.isComponentIdentifier)(componentName) ? componentName : null;
};
exports.getElementComponentNameFromSourceCode = getElementComponentNameFromSourceCode;
const makeElementDragData = ({ dimensions, displayName, slug, sourceCode, }) => {
    return {
        type: 'remotion-element',
        version: 1,
        element: {
            slug,
            displayName,
            sourceCode,
            dimensions,
        },
    };
};
exports.makeElementDragData = makeElementDragData;
const parseElementDragData = (value) => {
    try {
        const parsed = JSON.parse(value);
        if (!isRecord(parsed)) {
            return null;
        }
        if (parsed.type !== 'remotion-element' || parsed.version !== 1) {
            return null;
        }
        if (!isRecord(parsed.element)) {
            return null;
        }
        const { dimensions, displayName, slug, sourceCode } = parsed.element;
        if (!isSlug(slug) ||
            !isDisplayName(displayName) ||
            !isSourceCode(sourceCode) ||
            (0, exports.getElementComponentNameFromSourceCode)(sourceCode) === null ||
            (0, exports.makeElementFileNameFromSlug)(slug) === null ||
            !isDimensions(dimensions)) {
            return null;
        }
        return (0, exports.makeElementDragData)({
            slug,
            displayName,
            sourceCode,
            dimensions,
        });
    }
    catch (_a) {
        return null;
    }
};
exports.parseElementDragData = parseElementDragData;
