// @remotion/studio/dist/components/InspectorPanel/KeyframeEasingNavigator.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.KeyframeEasingNavigator = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const react_1 = require("react");
const remotion_1 = require("remotion");
const colors_1 = require("../../helpers/colors");
const InlineAction_1 = require("../InlineAction");
const InspectorPanelLayout_1 = require("../InspectorPanelLayout");
const TimelineKeyframeDiamondIcon_1 = require("../Timeline/TimelineKeyframeDiamondIcon");
const TimelineKeyframeEasingLine_1 = require("../Timeline/TimelineKeyframeEasingLine");
const TimelineSelection_1 = require("../Timeline/TimelineSelection");
const keyframe_easing_navigator_items_1 = require("./keyframe-easing-navigator-items");
const navigatorContainer = {
    alignItems: 'center',
    boxSizing: 'border-box',
    display: 'flex',
    gap: 8,
    justifyContent: 'space-between',
    padding: `8px ${InspectorPanelLayout_1.INSPECTOR_PANEL_HORIZONTAL_PADDING}px 6px`,
    width: '100%',
};
const itemStrip = {
    alignItems: 'center',
    display: 'flex',
    flex: 1,
    gap: 0,
    height: 24,
    justifyContent: 'center',
    minWidth: 0,
};
const itemButton = {
    alignItems: 'center',
    background: 'none',
    border: 'none',
    display: 'flex',
    height: 24,
    justifyContent: 'center',
    padding: 0,
};
const diamondSize = 12;
const easingLineDiamondOverlap = diamondSize / 2;
const emptyItem = {
    height: 24,
    width: diamondSize,
};
const easingItem = {
    height: 12,
    position: 'relative',
    width: 38,
};
const getEasingLineUnderlay = ({ extendLeft, extendRight, }) => {
    return {
        bottom: 0,
        left: extendLeft ? -easingLineDiamondOverlap : 0,
        pointerEvents: 'none',
        position: 'absolute',
        right: extendRight ? -easingLineDiamondOverlap : 0,
        top: 0,
    };
};
const chevronIcon = {
    display: 'block',
    height: 12,
    width: 8,
};
const disabledChevronColor = 'rgba(255, 255, 255, 0.25)';
const NavigationChevron = ({ color, direction }) => {
    const path = direction === 'left' ? 'M6 1L2 6L6 11' : 'M2 1L6 6L2 11';
    return (jsx_runtime_1.jsx("svg", { viewBox: "0 0 8 12", style: chevronIcon, children: jsx_runtime_1.jsx("path", { d: path, fill: "none", stroke: color, strokeLinecap: "round", strokeLinejoin: "round", strokeWidth: "1.5" }) }));
};
const TimelineNavigatorItem = ({ extendLineLeft, extendLineRight, item, onSelect, selected }) => {
    const onClick = (0, react_1.useCallback)(() => {
        if (item === null) {
            return;
        }
        onSelect(item);
    }, [item, onSelect]);
    if (item === null) {
        return jsx_runtime_1.jsx("div", { style: emptyItem });
    }
    const buttonStyle = {
        ...itemButton,
        position: 'relative',
        width: item.type === 'easing' ? 38 : diamondSize,
        zIndex: item.type === 'keyframe' ? 1 : 0,
    };
    const lineUnderlayStyle = getEasingLineUnderlay({
        extendLeft: extendLineLeft,
        extendRight: extendLineRight,
    });
    return (jsx_runtime_1.jsx("button", { type: "button", style: buttonStyle, onClick: onClick, title: item.type === 'keyframe'
            ? `Keyframe at frame ${item.selection.frame}`
            : `Easing from frame ${item.selection.fromFrame} to ${item.selection.toFrame}`, "aria-label": item.type === 'keyframe'
            ? `Select keyframe at frame ${item.selection.frame}`
            : `Select easing from frame ${item.selection.fromFrame} to ${item.selection.toFrame}`, children: item.type === 'keyframe' ? (jsx_runtime_1.jsx(TimelineKeyframeDiamondIcon_1.TimelineKeyframeDiamondIcon, { color: colors_1.LIGHT_TEXT, selected: selected, size: 12 })) : (jsx_runtime_1.jsx("div", { style: easingItem, children: jsx_runtime_1.jsx("div", { style: lineUnderlayStyle, children: jsx_runtime_1.jsx(TimelineKeyframeEasingLine_1.TimelineKeyframeEasingLineVisual, { selected: selected }) }) })) }));
};
const KeyframeEasingNavigator = ({ currentSelection, includeEasings, keyframes, nodePathInfo }) => {
    const { isSelected, selectItems } = (0, TimelineSelection_1.useTimelineSelection)();
    const setFrame = remotion_1.Internals.useTimelineSetFrame();
    const videoConfig = (0, remotion_1.useVideoConfig)();
    const items = (0, react_1.useMemo)(() => (0, keyframe_easing_navigator_items_1.getKeyframeEasingNavigatorItems)({
        includeEasings,
        keyframes,
        nodePathInfo,
    }), [includeEasings, keyframes, nodePathInfo]);
    const currentKey = (0, TimelineSelection_1.getTimelineSelectionKey)(currentSelection);
    const currentIndex = items.findIndex((item) => (0, TimelineSelection_1.getTimelineSelectionKey)(item.selection) === currentKey);
    const previousItem = currentIndex > 0 ? items[currentIndex - 1] : null;
    const currentItem = currentIndex === -1 ? null : items[currentIndex];
    const nextItem = currentIndex === -1 || currentIndex >= items.length - 1
        ? null
        : items[currentIndex + 1];
    const selectItem = (0, react_1.useCallback)((item) => {
        selectItems([item.selection], { reveal: true });
    }, [selectItems]);
    const seekToItem = (0, react_1.useCallback)((item) => {
        const frame = (0, keyframe_easing_navigator_items_1.getNavigatorItemPlayheadFrame)(item);
        setFrame((current) => {
            const next = { ...current, [videoConfig.id]: frame };
            remotion_1.Internals.persistCurrentFrame(next);
            return next;
        });
    }, [setFrame, videoConfig.id]);
    const selectPrevious = (0, react_1.useCallback)(() => {
        if (previousItem === null) {
            return;
        }
        seekToItem(previousItem);
        selectItem(previousItem);
    }, [previousItem, seekToItem, selectItem]);
    const selectNext = (0, react_1.useCallback)(() => {
        if (nextItem === null) {
            return;
        }
        seekToItem(nextItem);
        selectItem(nextItem);
    }, [nextItem, seekToItem, selectItem]);
    if (currentItem === null) {
        return null;
    }
    return (jsx_runtime_1.jsxs("div", { style: navigatorContainer, children: [jsx_runtime_1.jsx(InlineAction_1.InlineAction, { disabled: previousItem === null, onClick: selectPrevious, title: "Previous animation item", renderAction: (color) => (jsx_runtime_1.jsx(NavigationChevron, { color: previousItem === null ? disabledChevronColor : color, direction: "left" })) }), jsx_runtime_1.jsxs("div", { style: itemStrip, children: [
                    jsx_runtime_1.jsx(TimelineNavigatorItem, { extendLineLeft: false, extendLineRight: currentItem.type === 'keyframe', item: previousItem, onSelect: selectItem, selected: previousItem === null ? false : isSelected(previousItem.selection) }), jsx_runtime_1.jsx(TimelineNavigatorItem, { extendLineLeft: (previousItem === null || previousItem === void 0 ? void 0 : previousItem.type) === 'keyframe', extendLineRight: (nextItem === null || nextItem === void 0 ? void 0 : nextItem.type) === 'keyframe', item: currentItem, onSelect: selectItem, selected: isSelected(currentItem.selection) }), jsx_runtime_1.jsx(TimelineNavigatorItem, { extendLineLeft: currentItem.type === 'keyframe', extendLineRight: false, item: nextItem, onSelect: selectItem, selected: nextItem === null ? false : isSelected(nextItem.selection) })
                ] }), jsx_runtime_1.jsx(InlineAction_1.InlineAction, { disabled: nextItem === null, onClick: selectNext, title: "Next animation item", renderAction: (color) => (jsx_runtime_1.jsx(NavigationChevron, { color: nextItem === null ? disabledChevronColor : color, direction: "right" })) })] }));
};
exports.KeyframeEasingNavigator = KeyframeEasingNavigator;
