// @remotion/studio/dist/components/InspectorPanel/SequenceInspectorHeader.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SequenceInspectorHeaderWithDivider = exports.SequenceInspectorHeader = exports.useSequenceInspectorSourceLocation = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const react_1 = require("react");
const remotion_1 = require("remotion");
const InspectorSourceLocation_1 = require("../InspectorSourceLocation");
const use_open_sequence_in_editor_1 = require("../Timeline/use-open-sequence-in-editor");
const styles_1 = require("./styles");
const useSequenceInspectorSourceLocation = (sequence) => {
    const { canOpenInEditor, openInEditor, originalLocation } = (0, use_open_sequence_in_editor_1.useOpenSequenceInEditor)(sequence);
    const validatedLocation = (0, react_1.useMemo)(() => {
        var _a;
        if (!originalLocation ||
            !originalLocation.source ||
            !originalLocation.line) {
            return null;
        }
        return {
            source: originalLocation.source,
            line: originalLocation.line,
            column: (_a = originalLocation.column) !== null && _a !== void 0 ? _a : 0,
        };
    }, [originalLocation]);
    const openFileLocation = (0, react_1.useCallback)(() => {
        if (!canOpenInEditor) {
            return;
        }
        openInEditor();
    }, [canOpenInEditor, openInEditor]);
    return {
        canOpenInEditor,
        openFileLocation,
        validatedLocation,
    };
};
exports.useSequenceInspectorSourceLocation = useSequenceInspectorSourceLocation;
const useSequenceDisplayName = (track) => {
    var _a, _b;
    var _c;
    const { propStatuses } = (0, react_1.useContext)(remotion_1.Internals.VisualModePropStatusesContext);
    const nodePath = (_c = (_a = track.nodePathInfo) === null || _a === void 0 ? void 0 : _a.sequenceSubscriptionKey) !== null && _c !== void 0 ? _c : null;
    return (0, react_1.useMemo)(() => {
        var _a;
        const propStatusesForOverride = nodePath
            ? remotion_1.Internals.getPropStatusesCtx(propStatuses, nodePath)
            : undefined;
        const codeNameStatus = propStatusesForOverride === null || propStatusesForOverride === void 0 ? void 0 : propStatusesForOverride.name;
        const fallbackDisplayName = ((_a = track.sequence.controls) === null || _a === void 0 ? void 0 : _a.componentName) || '<Sequence>';
        if ((codeNameStatus === null || codeNameStatus === void 0 ? void 0 : codeNameStatus.status) === 'static' &&
            typeof codeNameStatus.codeValue === 'string') {
            return codeNameStatus.codeValue;
        }
        if ((codeNameStatus === null || codeNameStatus === void 0 ? void 0 : codeNameStatus.status) === 'static') {
            return fallbackDisplayName;
        }
        return track.sequence.displayName || fallbackDisplayName;
    }, [
        nodePath,
        propStatuses,
        (_b = track.sequence.controls) === null || _b === void 0 ? void 0 : _b.componentName,
        track.sequence.displayName,
    ]);
};
const SequenceInspectorHeader = ({ sourceLocation, track }) => {
    var _a;
    const sequenceDisplayName = useSequenceDisplayName(track);
    const { documentationLink } = track.sequence;
    const openDocumentationLink = (0, react_1.useCallback)(() => {
        if (!documentationLink) {
            return;
        }
        window.open(documentationLink, '_blank', 'noopener,noreferrer');
    }, [documentationLink]);
    const subtitleStyle = (0, react_1.useMemo)(() => {
        return {
            ...styles_1.sequenceHeaderSubtitle,
            cursor: documentationLink ? 'pointer' : 'default',
        };
    }, [documentationLink]);
    const componentName = (_a = track.sequence.controls) === null || _a === void 0 ? void 0 : _a.componentName;
    return (jsx_runtime_1.jsxs("div", { style: styles_1.sequenceHeader, children: [
            jsx_runtime_1.jsx("div", { style: styles_1.sequenceHeaderTitle, children: sequenceDisplayName }), documentationLink ? (jsx_runtime_1.jsx("button", { type: "button", style: subtitleStyle, title: "Open component docs", onClick: openDocumentationLink, children: componentName })) : (jsx_runtime_1.jsx("div", { style: subtitleStyle, children: componentName })), jsx_runtime_1.jsx(InspectorSourceLocation_1.InspectorSourceLocation, { location: sourceLocation.validatedLocation, canOpen: sourceLocation.canOpenInEditor, onOpen: sourceLocation.openFileLocation })
        ] }));
};
exports.SequenceInspectorHeader = SequenceInspectorHeader;
const SequenceInspectorHeaderWithDivider = ({ track }) => {
    const sourceLocation = (0, exports.useSequenceInspectorSourceLocation)(track.sequence);
    return (jsx_runtime_1.jsxs(jsx_runtime_1.Fragment, { children: [
            jsx_runtime_1.jsx(exports.SequenceInspectorHeader, { sourceLocation: sourceLocation, track: track }), jsx_runtime_1.jsx("div", { style: styles_1.sequenceHeaderDivider })
        ] }));
};
exports.SequenceInspectorHeaderWithDivider = SequenceInspectorHeaderWithDivider;
