// @remotion/studio/dist/components/Timeline/get-animation-item-selection-for-frame.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getAnimationItemSelectionForSourceFrame = void 0;
const getKeyframeSelection = ({ keyframe, keyframeDisplayOffset, nodePathInfo, }) => {
    return {
        type: 'keyframe',
        nodePathInfo,
        frame: keyframe.frame + keyframeDisplayOffset,
    };
};
const getAnimationItemSelectionForSourceFrame = ({ includeEasings, keyframeDisplayOffset, keyframes, nodePathInfo, sourceFrame, }) => {
    const firstKeyframe = keyframes[0];
    if (!firstKeyframe) {
        return null;
    }
    for (let i = 0; i < keyframes.length; i++) {
        const keyframe = keyframes[i];
        if (keyframe.frame === sourceFrame) {
            return getKeyframeSelection({
                keyframe,
                keyframeDisplayOffset,
                nodePathInfo,
            });
        }
        const nextKeyframe = keyframes[i + 1];
        if (!nextKeyframe) {
            continue;
        }
        if (sourceFrame > keyframe.frame && sourceFrame < nextKeyframe.frame) {
            if (!includeEasings) {
                return null;
            }
            return {
                type: 'easing',
                nodePathInfo,
                fromFrame: keyframe.frame + keyframeDisplayOffset,
                toFrame: nextKeyframe.frame + keyframeDisplayOffset,
                segmentIndex: i,
            };
        }
    }
    const lastKeyframe = keyframes[keyframes.length - 1];
    return getKeyframeSelection({
        keyframe: sourceFrame < firstKeyframe.frame ? firstKeyframe : lastKeyframe,
        keyframeDisplayOffset,
        nodePathInfo,
    });
};
exports.getAnimationItemSelectionForSourceFrame = getAnimationItemSelectionForSourceFrame;
